File: android_app/lib/ble/port_decoder_wiring.dart
import 'dart:async';
import 'dart:typed_data';
import '../utils/logger/logger.dart';
import '../protocol/crypto.dart';
import '../protocol/miot_tlv.dart';
import '../protocol/port_decoder.dart';
import 'android_connector.dart';

/// 将 BLE cmd_recv 通知连接到 PortDecoder
///
/// 在连接认证成功后调用此函数以启动端口数据解码。
/// 返回 [StreamSubscription] 以便调用方在断开连接时取消订阅。
Future<StreamSubscription<BleNotification>> wirePortDecoder(
  AndroidConnector connector,
) async {
  return connector.notifications
      .where((n) => n.channel == 'cmd_recv')
      .listen((n) async {
    try {
      final ciphertext = n.data;
      if (ciphertext.length < 3) return;

      // 使用加密通道解密
      final crypto = CryptoEngine.instance;
      if (!crypto.hasKeys) {
        AppLogger.instance.w('PortDecoder', 'No session keys, cannot decrypt');
        return;
      }

      final plaintext = await crypto.decrypt(ciphertext.sublist(2));
      if (plaintext == null || plaintext.isEmpty) {
        AppLogger.instance.w('PortDecoder', 'Decrypt failed');
        return;
      }

      // 尝试解析 TLV
      final tlv = MiotTlv.parse(Uint8List.fromList(plaintext));
      if (tlv == null) return;

      // 如果是端口推送 (opcode=0x02, siid=2)
      PortDecoder.dispatchFromTlvMap(tlv);
    } catch (e, st) {
      AppLogger.instance.e('PortDecoder', 'Decode error: $e', st);
    }
  }, onError: (e, st) {
    AppLogger.instance.e('PortDecoder', 'Stream error: $e', st);
  });
}

File: windows_app/src/CukTechController/ble/PortDecoderWiring.cs
using System;
using System.Linq;
using CukTechController.Protocol;
using CukTechController.Utils;

namespace CukTechController.Ble
{
    /// <summary>
    /// 将 BLE cmd_recv 通知连接到 PortDecoder
    /// 在认证成功后调用 WireUp() 启动端口数据解码
    /// </summary>
    public static class PortDecoderWiring
    {
        private static EventHandler<(string Channel, byte[] Data)>? _handler;

        /// <summary>
        /// 启动端口数据解码订阅
        /// </summary>
        public static void WireUp(WindowsConnector connector)
        {
            if (_handler != null)
            {
                connector.ValueReceived -= _handler;
            }

            _handler = (sender, e) =>
            {
                if (e.Channel != "cmd_recv") return;
                try
                {
                    var data = e.Data;
                    if (data.Length < 3) return;

                    var crypto = CryptoEngine.Instance;
                    if (!crypto.HasKeys)
                    {
                        AppLogger.Warn("PortDecoderWiring: No session keys, cannot decrypt");
                        return;
                    }

                    var ciphertext = data.Skip(2).ToArray();
                    var plaintext = crypto.Decrypt(ciphertext);
                    if (plaintext == null || plaintext.Length == 0)
                    {
                        AppLogger.Warn("PortDecoderWiring: Decrypt failed");
                        return;
                    }

                    var tlv = MiotTlv.Parse(plaintext);
                    if (tlv == null) return;

                    // 如果是端口推送 (opcode=0x02, siid=2)
                    PortDecoder.DispatchFromTlvMap(tlv);
                }
                catch (Exception ex)
                {
                    AppLogger.Error("PortDecoderWiring", $"Decode error: {ex.Message}", ex);
                }
            };

            connector.ValueReceived += _handler;
            AppLogger.Info("PortDecoderWiring", "Port decoder wired up on cmd_recv");
        }

        /// <summary>
        /// 停止端口数据解码订阅
        /// </summary>
        public static void TearDown(WindowsConnector connector)
        {
            if (_handler != null)
            {
                connector.ValueReceived -= _handler;
                _handler = null;
                AppLogger.Info("PortDecoderWiring", "Port decoder torn down");
            }
        }
    }
}
